e2etest/l2tests/assignment.go

```diff
@@ -145,6 +148,63 @@ var _ = ginkgo.Describe("IP Assignment", func() {
 					err := cs.CoreV1().Services(svc.Namespace).Delete(context.Background(), svc.Name, metav1.DeleteOptions{})
 					return err
 				}))
+
+		ginkgo.It("should preseve the same external ip after controller restart", func() {
+			const numOfRestarts = 5
+			resources := config.Resources{
+				Pools: []metallbv1beta1.IPAddressPool{
+					{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "assignment-controller-reset-test-pool",
+						},
+						Spec: metallbv1beta1.IPAddressPoolSpec{
+							Addresses: []string{"192.168.10.100/32"},
+						},
+					},
+				},
+			}
+			err := ConfigUpdater.Update(resources)
+			framework.ExpectNoError(err)
+
+			ginkgo.By("creating 4 LB services")
+			jig := e2eservice.NewTestJig(cs, f.Namespace.Name, "service-specific")
+			specificService, err := jig.CreateLoadBalancerService(30*time.Second, func(svc *v1.Service) {})
+			framework.ExpectNoError(err)
+			defer service.Delete(cs, specificService)
+			service.ValidateDesiredLB(specificService)
+
+			svc, err := jig.CreateLoadBalancerServiceWaitForClusterIPOnly(func(svc *v1.Service) { svc.ObjectMeta.Name = "service-b" })
+			framework.ExpectNoError(err)
+			defer service.Delete(cs, svc)
+
+			svc, err = jig.CreateLoadBalancerServiceWaitForClusterIPOnly(func(svc *v1.Service) { svc.ObjectMeta.Name = "service-c" })
+			framework.ExpectNoError(err)
+			defer service.Delete(cs, svc)
+
+			svc, err = jig.CreateLoadBalancerServiceWaitForClusterIPOnly(func(svc *v1.Service) { svc.ObjectMeta.Name = "service-d" })
+			framework.ExpectNoError(err)
+			defer service.Delete(cs, svc)
+
+			restartAndAssert := func() {
+				metallb.RestartController(cs)
+				gomega.Consistently(func() error {
+					specificService, err = cs.CoreV1().Services(specificService.Namespace).Get(context.TODO(), specificService.Name, metav1.GetOptions{})
```

Comment:
Sorry I wasn't clear in https://github.com/metallb/metallb/pull/2004#discussion_r1299981050

What I was suggesting there is to ensure that even a service without a specific IP maintains it's IP after the restart. In order to do that you'll have to loosen the cidr (24, for example) and assert that afer the restart the ip remains the same for at least one of the others

Reply:
I don't see how this will work, because when I loosen to /24 all of the services will get allocated and thus it wouldn't create the conditions for the bug to occur from the first place, where we have more LB services than available IPs and the controller assigns an used IP to a different service.

Reply:
The bug will happen anyway, because the controller will assign the ips in the order the services are processed.
So, before this fix, it was possible to have svcA, svcB and svcC to get ipA, ipB and ipC, restart the contorller and have them scrambled.

Reply:
Right but we want to ensure that a service that got an IP preserves it regardless it's asking or getting assigned. What we can do then is to have two ips and 4 services, and check that both services that got an IP keep their IPs.

Reply:
ok. done.